test/spmd/tensor/test_megatron_example.py

```diff
@@ -129,6 +187,44 @@ def test_simple_megatron_e2e(self):
         output_tp = model_tp(inp)
         self.assertEqual(output, output_tp)
 
+    @with_comms
+    def test_self_attn_megatron_e2e(self):
+        LR = 0.05
+        inp_size = [8, 12, 16]
+        torch.manual_seed(0)
+        inp = torch.rand(*inp_size, device=self.device_type)
+        torch.manual_seed(5)
+        model = TensorParallelMultiheadAttention(
+            16, 8, self.device_type, tp_size=4, add_bias_kv=True
+        )
+        torch.manual_seed(5)
+        model_tp = TensorParallelMultiheadAttention(
+            16, 8, self.device_type, tp_size=4, add_bias_kv=True
+        )
+        self.assertEqual(model.qkv.weight, model_tp.qkv.weight)
+        self.assertEqual(model.qkv.bias, model_tp.qkv.bias)
+        self.assertEqual(model.proj.weight, model_tp.proj.weight)
+        self.assertEqual(model.proj.bias, model_tp.proj.bias)
+        shard_self_attn(model_tp, self.device_type)
+
+        output = model(inp, inp, inp)
+        output_tp = model_tp(inp, inp, inp)
+        self.assertEqual(output, output_tp)
+
+        output.sum().backward()
+        output_tp.sum().backward()
+
+        optim = torch.optim.SGD(model.parameters(), lr=LR)
+        optim.step()
+        optim = torch.optim.SGD(model_tp.parameters(), lr=LR)
+        optim.step()
+
+        torch.manual_seed(3)
```

Comment:
Same comments as the previous example.

Reply:
This is creating the input for post-optimization.

Reply:
Yes, I am suggesting we use a training loop closer to how it is usually written.